File: tools/fetch-jguitar.js
#!/usr/bin/env node
/**
 * Fetch all guitar chord voicings from jguitar.com
 * Saves as a static JSON file for use in the app.
 *
 * Usage: node tools/fetch-jguitar.js
 * Output: data/jguitar-chords.json
 */

const fs = require('fs');
const path = require('path');

// All 12 root notes
const ROOTS = ['C', 'C#', 'D', 'Eb', 'E', 'F', 'F#', 'G', 'Ab', 'A', 'Bb', 'B'];

// Suffix → jguitar search suffix mapping
// jguitar uses specific naming conventions
const SUFFIXES = [
  { suffix: '',       search: '' },        // major
  { suffix: 'm',      search: 'm' },       // minor
  { suffix: 'dim',    search: 'dim' },
  { suffix: 'dim7',   search: 'dim7' },
  { suffix: 'sus2',   search: 'sus2' },
  { suffix: 'sus4',   search: 'sus4' },
  { suffix: '7sus4',  search: '7sus4' },
  { suffix: 'aug',    search: 'aug' },
  { suffix: '6',      search: '6' },
  { suffix: '69',     search: '6/9' },
  { suffix: '7',      search: '7' },
  { suffix: '7b5',    search: '7b5' },
  { suffix: 'aug7',   search: 'aug7' },
  { suffix: '9',      search: '9' },
  { suffix: '7b9',    search: '7b9' },
  { suffix: '7#9',    search: '7#9' },
  { suffix: '11',     search: '11' },
  { suffix: '13',     search: '13' },
  { suffix: 'maj7',   search: 'maj7' },
  { suffix: 'maj7#5', search: 'maj7#5' },
  { suffix: 'maj9',   search: 'maj9' },
  { suffix: 'maj11',  search: 'maj11' },
  { suffix: 'maj13',  search: 'maj13' },
  { suffix: 'm6',     search: 'm6' },
  { suffix: 'm69',    search: 'm6/9' },
  { suffix: 'm7',     search: 'm7' },
  { suffix: 'm7b5',   search: 'm7b5' },
  { suffix: 'm9',     search: 'm9' },
  { suffix: 'm11',    search: 'm11' },
  { suffix: 'mmaj7',  search: 'mMaj7' },
  { suffix: 'add9',   search: 'add9' },
  { suffix: 'madd9',  search: 'madd9' },
  { suffix: '5',      search: '5' },
  { suffix: '7sus2',  search: '7sus2' },
  { suffix: 'aug9',   search: 'aug9' },
  { suffix: '9#11',   search: '9#11' },
  { suffix: 'm13',    search: 'm13' },
  { suffix: 'mmaj9',  search: 'mMaj9' },
];

/**
 * Parse fret string from jguitar image URL.
 * e.g. "3,3,0,2,3,2" → [3,3,0,2,3,2]
 */
function parseJGuitarFrets(fretStr) {
  return fretStr.split(',').map(f => {
    const s = f.trim();
    if (s === 'x' || s === 'X') return -1;
    const n = parseInt(s, 10);
    return isNaN(n) ? -1 : n;
  });
}

/**
 * Convert absolute fret array to position format with baseFret.
 */
function toRelativeFrets(absFrets) {
  const fretted = absFrets.filter(f => f > 0);
  if (fretted.length === 0) {
    return { frets: [...absFrets], baseFret: 1 };
  }
  const minFret = Math.min(...fretted);
  const maxFret = Math.max(...fretted);
  const newBaseFret = maxFret <= 4 ? 1 : minFret;
  const relFrets = absFrets.map(f => {
    if (f <= 0) return f;
    return f - (newBaseFret - 1);
  });
  return { frets: relFrets, baseFret: newBaseFret };
}

/**
 * Convert absolute frets to position with barre detection.
 */
function absFretsToPosition(absFrets) {
  const rel = toRelativeFrets(absFrets);
  const barres = [];
  const fretCounts = {};
  rel.frets.forEach(f => {
    if (f > 0) fretCounts[f] = (fretCounts[f] || 0) + 1;
  });
  for (const [fret, count] of Object.entries(fretCounts)) {
    if (count >= 2) barres.push(parseInt(fret, 10));
  }
  return {
    frets: rel.frets,
    baseFret: rel.baseFret,
    barres,
  };
}

/**
 * Fetch chord voicings from jguitar.com
 */
async function fetchChord(chordName) {
  const url = `https://jguitar.com/chordsearch?chordsearch=${encodeURIComponent(chordName)}`;

  try {
    const res = await fetch(url, {
      signal: AbortSignal.timeout(15000),
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
      },
    });
    if (!res.ok) return null;

    const html = await res.text();

    const imgRegex = /images\/chordshape\/[^"]*?-([\dx%A-Fa-f,]+)\.png/gi;
    const positions = [];
    const seen = new Set();
    let match;

    while ((match = imgRegex.exec(html)) !== null) {
      const fretStr = decodeURIComponent(match[1]);
      if (seen.has(fretStr)) continue;
      seen.add(fretStr);

      const absFrets = parseJGuitarFrets(fretStr);
      if (absFrets.length !== 6) continue;
      if (absFrets.filter(f => f >= 0).length < 3) continue;

      positions.push(absFretsToPosition(absFrets));
    }

    return positions.length > 0 ? positions : null;
  } catch (e) {
    console.error(`  ✗ ${chordName}: ${e.message}`);
    return null;
  }
}

/**
 * Sleep helper
 */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Main: fetch all chords and save to JSON
 */
async function main() {
  const outputDir = path.join(__dirname, '..', 'data');
  if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

  const outputPath = path.join(outputDir, 'jguitar-chords.json');
  const result = {};
  let total = 0;
  let found = 0;
  let failed = 0;

  const totalChords = ROOTS.length * SUFFIXES.length;
  console.log(`Fetching ${totalChords} chord voicings from jguitar.com...\n`);

  for (const root of ROOTS) {
    for (const { suffix, search } of SUFFIXES) {
      const chordName = root + suffix;        // Our key (e.g. "C#m7")
      const searchName = root + search;       // jguitar search term
      total++;

      const progress = `[${total}/${totalChords}]`;
      process.stdout.write(`${progress} ${chordName}...`);

      const positions = await fetchChord(searchName);

      if (positions) {
        result[chordName] = positions;
        found++;
        console.log(` ✓ ${positions.length} voicings`);
      } else {
        failed++;
        console.log(` ✗ not found`);
      }

      // Rate limiting: 300ms between requests
      await sleep(300);
    }
  }

  // Save result
  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2), 'utf-8');

  const fileSize = (fs.statSync(outputPath).size / 1024).toFixed(1);
  console.log(`\n========================================`);
  console.log(`Total: ${total} | Found: ${found} | Failed: ${failed}`);
  console.log(`Saved to: ${outputPath} (${fileSize} KB)`);
  console.log(`========================================`);
}

main().catch(console.error);

File: tools/fetch-jguitar.py
#!/usr/bin/env python3
"""
Fetch all guitar chord voicings from jguitar.com
Saves as a static JSON file for use in the app.

Usage: python3 tools/fetch-jguitar.py
Output: data/jguitar-chords.json
"""

import json
import os
import re
import sys
import time
import urllib.request
import urllib.parse
import urllib.error

# All 12 root notes
ROOTS = ['C', 'C#', 'D', 'Eb', 'E', 'F', 'F#', 'G', 'Ab', 'A', 'Bb', 'B']

# Suffix → jguitar search suffix mapping
SUFFIXES = [
    ('', ''),           # major
    ('m', 'm'),         # minor
    ('dim', 'dim'),
    ('dim7', 'dim7'),
    ('sus2', 'sus2'),
    ('sus4', 'sus4'),
    ('7sus4', '7sus4'),
    ('aug', 'aug'),
    ('6', '6'),
    ('69', '6/9'),
    ('7', '7'),
    ('7b5', '7b5'),
    ('aug7', 'aug7'),
    ('9', '9'),
    ('7b9', '7b9'),
    ('7#9', '7#9'),
    ('11', '11'),
    ('13', '13'),
    ('maj7', 'maj7'),
    ('maj7#5', 'maj7#5'),
    ('maj9', 'maj9'),
    ('maj11', 'maj11'),
    ('maj13', 'maj13'),
    ('m6', 'm6'),
    ('m69', 'm6/9'),
    ('m7', 'm7'),
    ('m7b5', 'm7b5'),
    ('m9', 'm9'),
    ('m11', 'm11'),
    ('mmaj7', 'mMaj7'),
    ('add9', 'add9'),
    ('madd9', 'madd9'),
    ('5', '5'),
    ('7sus2', '7sus2'),
    ('aug9', 'aug9'),
    ('9#11', '9#11'),
    ('m13', 'm13'),
    ('mmaj9', 'mMaj9'),
]


def parse_frets(fret_str):
    """Parse fret string like '3,3,0,2,3,2' → [3,3,0,2,3,2]"""
    result = []
    for f in fret_str.split(','):
        s = f.strip()
        if s.lower() == 'x':
            result.append(-1)
        else:
            try:
                result.append(int(s))
            except ValueError:
                result.append(-1)
    return result


def to_relative_frets(abs_frets):
    """Convert absolute frets to relative with baseFret."""
    fretted = [f for f in abs_frets if f > 0]
    if not fretted:
        return list(abs_frets), 1

    min_fret = min(fretted)
    max_fret = max(fretted)
    new_base = 1 if max_fret <= 4 else min_fret

    rel_frets = []
    for f in abs_frets:
        if f <= 0:
            rel_frets.append(f)
        else:
            rel_frets.append(f - (new_base - 1))
    return rel_frets, new_base


def abs_frets_to_position(abs_frets):
    """Convert absolute frets to position dict with barre detection."""
    rel_frets, base_fret = to_relative_frets(abs_frets)

    # Detect barres
    fret_counts = {}
    for f in rel_frets:
        if f > 0:
            fret_counts[f] = fret_counts.get(f, 0) + 1
    barres = [f for f, count in fret_counts.items() if count >= 2]

    return {
        'frets': rel_frets,
        'baseFret': base_fret,
        'barres': sorted(barres),
    }


def fetch_chord(chord_name):
    """Fetch chord voicings from jguitar.com"""
    url = f'https://jguitar.com/chordsearch?chordsearch={urllib.parse.quote(chord_name)}'

    req = urllib.request.Request(url, headers={
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
    })

    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            html = response.read().decode('utf-8', errors='replace')
    except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError, Exception) as e:
        return None, str(e)

    # Extract fret positions from image URLs
    img_regex = re.compile(r'images/chordshape/[^"]*?-([\dx%A-Fa-f,]+)\.png', re.IGNORECASE)
    positions = []
    seen = set()

    for match in img_regex.finditer(html):
        fret_str = urllib.parse.unquote(match.group(1))
        if fret_str in seen:
            continue
        seen.add(fret_str)

        abs_frets = parse_frets(fret_str)
        # Must have 6 values for guitar
        if len(abs_frets) != 6:
            continue
        # Must have at least 3 sounding strings
        if sum(1 for f in abs_frets if f >= 0) < 3:
            continue

        positions.append(abs_frets_to_position(abs_frets))

    return positions if positions else None, None


def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_dir = os.path.dirname(script_dir)
    output_dir = os.path.join(project_dir, 'data')
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, 'jguitar-chords.json')

    result = {}
    total = 0
    found = 0
    failed = 0
    total_chords = len(ROOTS) * len(SUFFIXES)

    print(f'Fetching {total_chords} chord voicings from jguitar.com...\n')

    for root in ROOTS:
        for suffix, search in SUFFIXES:
            chord_name = root + suffix       # Our key (e.g. "C#m7")
            search_name = root + search      # jguitar search term
            total += 1

            progress = f'[{total}/{total_chords}]'
            sys.stdout.write(f'{progress} {chord_name}...')
            sys.stdout.flush()

            positions, error = fetch_chord(search_name)

            if positions:
                result[chord_name] = positions
                found += 1
                print(f' \u2713 {len(positions)} voicings')
            else:
                failed += 1
                if error:
                    print(f' \u2717 error: {error[:50]}')
                else:
                    print(f' \u2717 not found')

            # Rate limiting: 300ms between requests
            time.sleep(0.3)

    # Save result
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(result, f, ensure_ascii=False)

    file_size = os.path.getsize(output_path) / 1024
    print(f'\n========================================')
    print(f'Total: {total} | Found: {found} | Failed: {failed}')
    print(f'Saved to: {output_path} ({file_size:.1f} KB)')
    print(f'========================================')


if __name__ == '__main__':
    main()
